lib/transport/transport-tls.c

```diff
@@ -24,23 +24,120 @@
  */
 
 #include "transport/transport-tls.h"
-#include "transport/transport-socket.h"
+#include "transport/transport-adapter.h"
 
 #include "messages.h"
 
 #include <openssl/ssl.h>
 #include <openssl/err.h>
 #include <errno.h>
 
-const gchar *TLS_TRANSPORT_NAME = "tls";
-
 typedef struct _LogTransportTLS
 {
-  LogTransportSocket super;
+  LogTransportAdapter super;
   TLSSession *tls_session;
   gboolean sending_shutdown;
 } LogTransportTLS;
 
+const gchar *TLS_TRANSPORT_NAME = "tls";
+
+static int
+_BIO_transport_write(BIO *bio, const char *buf, size_t buflen, size_t *written_bytes)
+{
+  LogTransportTLS *transport = BIO_get_data(bio);
+  gssize ret;
+
+  ret = log_transport_adapter_write_method(&transport->super.super, (gpointer) buf, buflen);
+  BIO_clear_retry_flags(bio);
+
+  if (ret < 0)
+    {
+      *written_bytes = 0;
+      if (errno == EAGAIN)
+        BIO_set_retry_write(bio);
+      return -1;
+    }
+  *written_bytes = ret;
+  return 1;
+}
+
+static int
+_BIO_transport_read(BIO *bio, char *buf, size_t buflen, size_t *read_bytes)
+{
+  LogTransportTLS *transport = BIO_get_data(bio);
+  gssize ret;
+
+  ret = log_transport_adapter_read_method(&transport->super.super, buf, buflen, NULL);
+  if (ret < 0)
+    {
+      *read_bytes = 0;
+      if (errno == EAGAIN)
+        BIO_set_retry_read(bio);
+      return -1;
+    }
+  *read_bytes = ret;
+  return 1;
+}
+
+static long
+_BIO_transport_ctrl(BIO *bio, int cmd, long num, void *ptr)
+{
+  long ret = 1;
+
+  switch (cmd)
+    {
+    case BIO_CTRL_GET_CLOSE:
+      ret = BIO_get_shutdown(bio);
+      break;
+    case BIO_CTRL_SET_CLOSE:
+      BIO_set_shutdown(bio, (int)num);
+      break;
+    case BIO_CTRL_DUP:
+    case BIO_CTRL_FLUSH:
+      ret = 1;
+      break;
+    case BIO_CTRL_RESET:
+    case BIO_C_FILE_SEEK:
+    case BIO_C_FILE_TELL:
+    case BIO_CTRL_INFO:
+    case BIO_C_SET_FD:
+    case BIO_C_GET_FD:
+    case BIO_CTRL_PENDING:
+    case BIO_CTRL_WPENDING:
+    default:
+      ret = 0;
+      break;
+    }
+
+  return ret;
+}
+
+BIO_METHOD *
+BIO_s_transport(void)
+{
+  static BIO_METHOD *meth = NULL;
+
+  if (meth)
+    return meth;
+
+  meth = BIO_meth_new(BIO_TYPE_NONE, "LogTransportBIO");
```

Comment:
This seems to be leaked. Shouldn't we call BIO_meth_free() through our apphooks before shutdown?

Reply:
added a transport-globals module and a global_init/deinit which in turn allocates and deallocates this through
apphook.